mesa/agent.py

```diff
@@ -356,6 +356,80 @@ def random(self) -> Random:
         """
         return self.model.random
 
+    def apply(
+        self, func: Callable, axis: str = "agent", args=(), result_type=None, **kwargs
+    ) -> list[Any] | Any:
+        """
+        Apply a function to all agents in the AgentSet either to each agent individually or to the entire agentset.
+
+        Args:
+            func (Callable): The function to apply to each individual agent or the entire agentset
+            axis (str): {'agent', 'agetset'} The axis along which to apply the function.
+
+                         * 'agent' means apply the function to each agent.
+                         * 'agentset' means apply the function to the entire agentset.
+
+            args (list or tuple):Positional arguments to pass to the function.
+            kwargs (dict): Additional keyword arguments to pass as keywords arguments to the function.
+
+        Returns:
+            the result of applying the function along the specified axis. In case of ``axis='agent'``, it will be a list
+            with the return of `func` for each agent. In case of ``axis='agentset'``, it is the return of ``func``.
+
+        Notes:
+            To maintain method chaining in case of axis=agentset, func should return an AgentSet
+
+        """
+        if axis == "agent":
+            # TODO:: add a results_type to make it trivial to return a dataframe with agent.id and func results?
+            # TODO:: this is a good idea, but tricky because you don't know all column names
+            return [func(agent, *args, **kwargs) for agent in self]
+        elif axis == "agentset":
+            return func(self, *args, **kwargs)
+        else:
+            raise ValueError(f"axis should be `agent` or `agentset` not {axis}")
+
+    def group_by(
```

Comment:
This is the [naming choice of Polars](https://docs.pola.rs/api/python/stable/reference/dataframe/api/polars.DataFrame.group_by.html) as well.